File: 1. Arrays & Strings/4_most_freq_char.java
import java.util.HashMap;

class Source {
    public static char mostFrequentChar(String s) {

        // build frequency map
        HashMap<Character, Integer> map = new HashMap<>();

        for (char c : s.toCharArray()) {
            if (map.get(c) == null) {
                map.put(c, 1); // 1st occ
            } else {
                int currentOcc = map.get(c);
                map.put(c, currentOcc + 1);
            }
        }

        char mostFreq = '0';
        int currentMaxOcc = 0;

        /**
         * find most frequent char
         * For the tiebreaking logic the original String
         * has the "keyset" in order of insertion
         */
        for (char c : s.toCharArray()) {
            if (map.get(c) > currentMaxOcc) {
                currentMaxOcc = map.get(c);
                mostFreq = c;
            }
        }

        return mostFreq;
    }
}

File: 1. Arrays & Strings/4_most_freq_char/4_most_freq_char.py
def most_frequent_char(s):
    # build frequency map
    char_map = {}

    for char in s:
        if char not in char_map:
            char_map[char] = 1  # 1st occ
        else:
            char_map[char] += 1
    
    most_freq = '0'
    current_max_occ = 0

    # find most frequent char
    for char in s:
        if char_map[char] > current_max_occ:
            current_max_occ = char_map[char]
            most_freq = char
    
    return most_freq
